### PR TITLE
Use only public api for ReactDOMEventListener-test.js

### PR DESCRIPTION
**Pull request by enapupe**
Related to  #11299 – Express more tests via public API
Hey! After a couple hours fighting with this one I think I got the idea.. 
Could you guys check if I'm in the right direction?

BTW, I lost quite some time until I realized this line: https://github.com/facebook/react/blob/master/packages/react-dom/src/__tests__/ReactDOMEventListener-test.js#L32  was messing up with `ReactTestUtils.SimulateNative`...  Funny, that's probably the reason why we are using just the public api 😅

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 1]
```diff
@@ -33,54 +33,53 @@ describe('ReactDOMEventListener', () => {
...
-      view: window,
-      path: [otherNode, otherNode],
-    });
-    expect(handleTopLevel.mock.calls.length).toBe(1);
+    var component = ReactDOM.render(
+      <div onMouseOver={mock} />,
+      document.createElement('div'),
+    );
+
+    ReactDOM.findDOMNode(component).appendChild(otherNode);
```
Comment (enapupe) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: I'm not sure about this first test, it seems to have lost his initial purpose.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 2]
```diff
@@ -33,54 +33,53 @@ describe('ReactDOMEventListener', () => {
...
+    ReactTestUtils.SimulateNative.mouseOver(otherNode);
```
Comment (sebmarkbage) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: For these purposes, I'd like to treat `SimulateNative` (and all the ReactTestUtils really as an internal API. The reason for that is that they also reach into internals indirectly. They actually bypass some of the implementation details that would make this example not work with normal React.

In normal React this wouldn't work because the container div is not in the document and therefore mouse moves would never be dispatched on top of it. React can't pick it up because it uses event listeners on the document to do it.

Instead, we should insert the container div into the document, dispatch a native event and then remove the div from the document. Like this:
https://github.com/facebook/react/blob/master/packages/react-dom/src/__tests__/ReactDOMInput-test.js#L55-L59

Comment (enapupe) [reply packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Hey! I did the requested changes, could you take a look at it?
I'm still not sure if I'm the first test is on spot because before it tested two unrelated components and now they have a parent-child relationship. 

Also while we are at this file, I'm not sure how to rewrite `should not get confused by disappearing elements`.. I think I understand the thinking behind it but I'm not sure if it's possible to accomplish the same result without mocking `handleTopLevel`. I'd have to unmount a component in the middle of the event chain, which seems impossible.

Thanks!

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 3]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
-      relatedTarget: ReactDOM.findDOMNode(component),
-      view: window,
-      path: [otherNode, otherNode],
-    });
-    expect(handleTopLevel.mock.calls.length).toBe(1);
+    var mock = jest.fn();
+
+    var container = document.createElement('div');
+    var stub = ReactDOM.render(<div onMouseOver={mock} />, container);
+    var node = ReactDOM.findDOMNode(stub);
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: There's no "stubs" anymore, you can remove these `findDOMNode()` calls. Return value of `ReactDOM.render(<div />)` is the node itself.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 4]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
+    var otherComponent = document.createElement('h1');
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Let's not call nodes "components". I understand this naming already existed but let's keep it clear in new code. We call DOM nodes "nodes" in newer code (or "containers" when we render into them). So `otherNode` was a good naming.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 5]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
+    node.appendChild(otherComponent);
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Why is it appended as a child to `node`? I don't see it in the original test. My impression (based on the test name) is that the test specifically wants to test it still works if the `otherComponent` is *outside* the React tree. Whereas this is putting it *inside*.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 6]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
+    document.body.appendChild(container);
+
+    var nativeEvent = document.createEvent('Event');
+    nativeEvent.initEvent('mouseover', true, true);
+    otherComponent.dispatchEvent(nativeEvent);
+
+    expect(mock).toBeCalled();
+    document.body.removeChild(container);
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: I think the right way to write this test would be something like:

```js
    var mock = jest.fn();

    var container = document.createElement('div');
    var node = ReactDOM.render(<div onMouseEnter={mock} />, container);
    var otherNode = document.createElement('h1');
    document.body.appendChild(container);
    document.body.appendChild(otherNode);

    var nativeEvent = document.createEvent('MouseEvent');
    nativeEvent.initMouseEvent('mouseout', true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, node);
    otherNode.dispatchEvent(nativeEvent);
    expect(mock).toBeCalled();
```

This way we verify that if an event originates outside of React tree (on `otherNode`) it still fires an event inside a React tree.

I had to use a different event but that doesn't matter for this test.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 7]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
       ReactDOM.findDOMNode(parentControl).appendChild(childContainer);
+      document.body.appendChild(parentContainer);
 
-      var callback = ReactDOMEventListener.dispatchEvent.bind(null, 'test');
-      callback({
-        target: ReactDOM.findDOMNode(childControl),
-      });
+      var nativeEvent = document.createEvent('Event');
+      nativeEvent.initEvent('mouseout', true, true);
+      ReactDOM.findDOMNode(childControl).dispatchEvent(nativeEvent);
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: No `findDOMNode` please.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 8]
```diff
@@ -25,62 +25,67 @@ describe('ReactDOMEventListener', () => {
...
+
       var childContainer = document.createElement('div');
-      var childControl = <div>Child</div>;
+      var childControl = <div onMouseOut={onMouseOut}>Child</div>;
       var parentContainer = document.createElement('div');
-      var parentControl = <div>Parent</div>;
+      var parentControl = <div onMouseOut={onMouseOut}>div</div>;
       childControl = ReactDOM.render(childControl, childContainer);
       parentControl = ReactDOM.render(parentControl, parentContainer);
       ReactDOM.findDOMNode(parentControl).appendChild(childContainer);
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Here, too, we can remove all `findDOMNode`s.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 9]
```diff
@@ -90,22 +95,25 @@ describe('ReactDOMEventListener', () => {
...
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Same re: findDOMNode.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 10]
```diff
@@ -187,28 +195,27 @@ describe('ReactDOMEventListener', () => {
...
 
       render() {
-        var inner = <div ref="inner">Inner</div>;
+        var inner = <div onMouseOut={onMouseOut} ref="inner">Inner</div>;
         return <div><div id="outer">{inner}</div></div>;
       }
     }
 
     var instance = ReactTestUtils.renderIntoDocument(<Wrapper />);
+    ReactTestUtils.SimulateNative.mouseOut(instance.getInner());
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Let's rewrite this one not to use `SimulateNative` either. Can we?

**Comment by enapupe** [comment]
Thanks for the feedback!
The changes were addressed but there are still 2 tests left to go. I'm not sure how to proceed here because both tests change `handleTopLevel` behaviour. 
One seems to test if `ReactDOMEventListener` can handle an element being unmounted in the middle of an event chain. 
The other I'm not even sure if I understand, it mentions two different roots enqueuing events, but the current test only dispatches one event so I'm not really sure what's that about. THanks!

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 11]
```diff
@@ -26,86 +25,99 @@ describe('ReactDOMEventListener', () => {
...
-      toElement: ReactDOM.findDOMNode(component),
-      relatedTarget: ReactDOM.findDOMNode(component),
-      view: window,
-      path: [otherNode, otherNode],
-    });
-    expect(handleTopLevel.mock.calls.length).toBe(1);
+    document.body.appendChild(container);
+    document.body.appendChild(otherNode);
+
+    var nativeEvent = document.createEvent('MouseEvent');
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: I saw another PR using `new MouseEvent(type, args)`, could you check if that one's going to be shorter?

Comment (enapupe) [reply packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Yeah it seems

**Comment by gaearon** [comment]
>One seems to test if ReactDOMEventListener can handle an element being unmounted in the middle of an event chain.

I think it just wants to test that if you `unmountComponentAtNode()` during a React event (e.g. in child component's handler), the parent still receives the event too. Does that seem to make sense? Can you try to write a test for it? It might also help to view the blame for these lines, find the code that “fixed” this case (when this code was introduced), and then “undo” the fix to see if your test catches the regression. This would tell you if you’re testing for the same thing it was originally written for or not.

>The other I'm not even sure if I understand, it mentions two different roots enqueuing events, but the current test only dispatches one event so I'm not really sure what's that about. THanks!

React batches updates occurring inside the same browser event. Means it applies all changes when it exists the top-level browser event handler (instead of when it exits component handler). This test verifies this behavior is true even if two components live in separate React trees (but in the same document).

In other words it wants to verify that if you do

```js
var mock = jest.fn();
```

and then, inside child event handler:

```js
ReactDOM.render(<div>1</div>, childContainer);
mock(childNode.textContent);
```

and inside parent event handler:

```js
ReactDOM.render(<div>2</div>, childContainer);
mock(childNode.textContent);
```

Both calls to `mock` should've been called with `'Child'` because **both updates are batched together and neither one is applied yet inside React event handlers**. You should, however, check that when after dispatching is done, `childNode.textContent` is now set to `'2'`. (Now that updates are flushed.)

**Comment by gaearon** [comment]
Looks like a test is still failing. Would you like to continue working on this?

**Comment by enapupe** [comment]
Yeah there's just one missing.
I'll try to do it until tomorrow, otherwise I'll be traveling for a few days and will be back only after the Nov 5th, feel free to take over if you must.

**Comment by gaearon** [comment]
Thanks, sounds good!

**Comment by enapupe** [comment]
So, I think I was able to reproduce the same test case but I couldn't find the code that introduced that fix you mentioned, so confirmation is still necessary. The blame for that lines seem to be sort of detached from the git tree..

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 12]
```diff
@@ -9,206 +9,201 @@
...
-        topLevelType,
-        topLevelTarget,
-        topLevelTargetID,
-        nativeEvent,
-      ) {
-        ReactDOM.render(
-          <div>{topLevelTarget === childNode ? '1' : '2'}</div>,
-          childContainer,
-        );
-        // Since we're batching, neither update should yet have gone through.
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Comments like this are valuable because they're not obvious.

**Comment by gaearon** [comment]
This looks great! Thanks. I made a few small tweaks.
